hal/src/thumbv7em/clock/v2/xosc.rs

```diff
@@ -311,37 +267,83 @@ impl<X: XoscNum> XoscToken<X> {
 pub type StartUp = STARTUP_A;
 
 /// Type alias for Xosc Input pin
-pub type XIn<X> = Pin<<X as XoscNum>::XIn, FloatingDisabled>;
+pub type XIn<X> = Pin<<X as XoscId>::XIn, FloatingDisabled>;
 
 /// Type alias for Xosc Output pin
-pub type XOut<X> = Pin<<X as XoscNum>::XOut, FloatingDisabled>;
+pub type XOut<X> = Pin<<X as XoscId>::XOut, FloatingDisabled>;
 
 //==============================================================================
 // Mode structure for Xosc
 //==============================================================================
 
-/// Trait that defines a mode [`Xosc`] is operating in
-pub trait Mode: Sealed {}
+mod private {
+    use super::{CrystalCurrent, Sealed, XOut, XoscId, XoscToken};
 
-/// Struct representing a clock mode for [`Xosc`]
+    /// Private trait to restrict access to the `configure` function
+    pub trait Mode<X: XoscId>: Sealed {
+        const FROM_CRYSTAL: bool;
+        type Config;
+        fn configure(token: &mut XoscToken<X>, config: &Self::Config);
```

Comment:
Hmm. That's another approach. But in that case, shouldn't the corresponding trait be `unsafe` to implement? That way anyone editing the code will know that implementing types should **not** be constructable by users.

There's also another problem with that approach. In the current implementation, specifying an `Xosc` in `CrystalMode` requires redundant types, i.e. `Xosc<XoscId0, CrystalMode<XoscId0>>`. You have to specify `XoscId0` *twice*. I refactored the `Mode` trait to get rid of that redundancy. That's why I added the type parameter and associated `Config` type. With the new implementation, you can just write `Xosc<XoscId0, CrystalMode>`. I don't think you can get rid of that redundancy with the `Self` approach. In that case, the `CrystalMode` type would have to be generic.

However, I think your approach could still work if I rearrange things slightly. I just have to make `CrystalConfig` public, but make it impossible to construct by users.

Although, on second thought, that might not work for `type Config = ();`. I would probably have to create some dummy type that can be instantiated by the `xosc` module but not by users. I'll have to think more about this.

Reply:
Maybe `Xosc<CrystalMode<XoscId0>>` and `Xosc<ClockMode<XoscId0>>` with associated `type Id: XoscId` constant for `Mode` trait is the way?
```rust
pub trait Mode: Sealed {
    type Id: XoscId;
    fn configure(&self, token: &mut XoscToken<Self::Id>);
}
pub struct ClockMode<X: XoscId>(PhantomData<X>)
impl<X: XoscId> Mode for ClockMode<X> {
    type Id: XoscId = X;
    fn configure(&self, token: &mut XoscToken<Self::Id>) {
    ....
    }
}
pub struct CrystalMode<X: XoscId>{
    xout: XOut<X>,
    current: CrystalCurrent,
    amplitude_loop_control: bool,
    low_buf_gain: bool,
}
impl<X: XoscId> Mode for CrystalMode<X> {
    type Id: XoscId = X;
    fn configure(&self, token: &mut XoscToken<Self::Id>) {
    ....
    }
}
```

Then you have access to `Id` everywhere you need it, including `Xosc` itself through `Mode::Id`. And notion of `Config` can be then flattened back to `Mode`. So all the fields `CrystalConfig` is holding can be moved back to `CrystalMode`. Maybe `Mode` should `unsafe` to implement with the comment that structs implementing it must be inconstructible in user code.

Reply:
Let me back up for a moment and explain in more detail why I was trying to refactor `xosc` (and `dpll`) in the first place.

Here is a list of the existing clock types and their generic parameters:
- `AhbClk<A> where A: AhbId`
- `ApbClk<A> where A: ApbId`
- `Gclk<G, I> where G: GclkId, I: GclkSourceId`
- `GclkIn<G, I> where G: GclkId, I: GclkIo<G>`
- `GclkOut<G, I> where G: GclkId, I: GclkIo<G>`
- `Pclk<P, I> where P: PclkId, I: PclkSourceId`
- `Dfll<M> where M: Mode`
- `Dpll<D, M> where D: DpllId, M: SrcMode<D>`
- `Xosc<X, M> where X: XoscId, M: Mode`
- `OscUlpBase`
- `OscUlp1k`
- `OscUlp32k`
- `XoscBase<M> where M: Mode`
- `Xosc32k`
- `Xosc1k`
- `RtcOsc<I> where I: RtcSourceId`

The pattern here is clear:
- Where a clock has more than one instance (e.g. `Gclk`), the first type parameter is an identifier for the specific instance (e.g. `Gclk5Id`). If there is only one instance (e.g. `Dfll`), there is no corresponding type parameter.
    - Note, this lets us easily create type aliases of the form `pub type Gclk5<I> = Gclk<Gclk5Id, I>;`
- Where a clock needs some amount of type-level configuration, it takes a second type parameter, usually indicating something about the source or output of the clock.

In general, the second type parameter is populated by a *single* type that *does not itself* have any type parameters. In fact, that was one of the primary reasons to create the `Id` types in the first place. However, that pattern is currently broken in two places:
- The second type parameter for `Xosc<X, M>` can be `CrystalMode<X> where X: XoscId`
- The second type parameter for `Dpll<D, M>` is one of three `*Driven` structs that each takes two type parameters. The first of each `*Driven` struct is always `D: DpllId`, and the second is one of `PclkSourceId == GclkId`, `Xosc32kId`, or `X: XoscId`.

Notice how the concrete types used for the second type parameter always *repeat* the first type parameter, i.e. `D` for `Dpll` and `X` for `Xosc`. This is redundant, and it breaks the otherwise consistent pattern throughout the rest of the module.

My aim for refactoring both the `xosc` and `dpll` modules was to remove the redundancy and make the entire `clock` module uniform, where the second type parameter is always a *single* type with no type parameters of its own.

Your latest proposal would actual break the pattern further. It would define `Xosc` as `Xosc<M> where M: Mode<X> where X: XoscId`. If we took that approach, we would no longer be able to define `Xosc0` and `Xosc1` aliases.

Moreover, if we want a consistent pattern, we need to *remove* the type parameter from `CrystalMode`, pushing it onto the trait, i.e. `Mode<X>`. That's what my initial proposal does.

I agree that the first draft of my implementation was not as clear as it could be, but I think that can be improved. I'll try to come up with an implementation you like better. But I would really like to make the pattern consistent across all of the clocks. I think that will help tremendously withe comprehensibility. Eventually, you should be able to write:
- `Dpll0<Gclk5Id>` for a `Pclk` driven DPLL
- `Dpll0<Xosc32kId>` for an XOSC32K driven DPLL
- `Dpll0<Xosc1Id>` for an XOSC driven DPLL
- `Xosc1<CrystalMode>` for a crystal driven XOSC

I think this structure is really uniform and worth bending the implementation to accommodate.

Reply:
I've now implemented all of these changes. I think there is only one situation where I couldn't remove a type parameter from the configuration type. `dfll::OpenLoop` can be a single type, without type parameters. But `dfll::ClosedLoop<G>` *must* have a type parameter, because it parameterizes which `Pclk<DfllId, G>` was used. The `G` type parameter only applies in `ClosedLoop` `Mode`, so that is the natural way to model it.

This case is different than the `Xosc` and `Dpll` cases above, because the `G` type parameter is not redundant, like the `X` and `D` type parameters were.